File: InMotionGIT.Common - Copy/Helpers/AuthenticationHandler.vb
Imports VB = Microsoft.VisualBasic

Namespace Helpers

    Public Class AuthenticationHandler

#Region "Random Password Methods"

        Public Shared Function CreateAdministratorPassword(requiredPasswordLength As Integer, nonAlphanumericCharacters As Integer) As String
            Dim newPassword As String = String.Empty
            Dim characters As String = String.Empty
            Dim passwordLength As Integer = requiredPasswordLength

            If nonAlphanumericCharacters > 0 Then
                characters = CreateRandomAlphanumericCharacters(nonAlphanumericCharacters)
                passwordLength = requiredPasswordLength - characters.Length
            End If

            For index As Integer = 0 To passwordLength - 1
                Select Case index
                    Case 0
                        newPassword += "a"
                    Case 1
                        newPassword += "d"
                    Case 2
                        newPassword += "m"
                    Case 3
                        newPassword += "i"
                    Case 4
                        newPassword += "n"
                    Case Else
                        newPassword += "1"
                End Select
            Next

            If characters.Length > 0 Then
                newPassword += characters
            End If

            Return newPassword
        End Function

        Public Shared Function CreateRandomPassword(requiredPasswordLength As Integer, nonAlphanumericCharacters As Integer) As String
            Dim newPassword As String = String.Empty
            Dim characters As String = String.Empty
            Dim passwordLength As Integer = requiredPasswordLength
            Const allowedChars As String = "abcdefghijkmnopqrstuvwxyzABCDEFGHJKLMNOPQRSTUVWXYZ0123456789"

            If nonAlphanumericCharacters > 0 Then
                characters = CreateRandomAlphanumericCharacters(nonAlphanumericCharacters)
                passwordLength = requiredPasswordLength - characters.Length
            End If

            Dim newRandom As New Random()

            For index As Integer = 0 To passwordLength - 1
                newPassword += allowedChars(newRandom.[Next](0, allowedChars.Length))
            Next

            If characters.Length > 0 Then
                newPassword += characters
            End If

            Return newPassword
        End Function

        Public Shared Function IsCorrectPasswordBySettings(currentPassword As String, requiredPasswordLength As Integer, nonAlphanumericCharacters As Integer) As Boolean
            Dim result As Boolean = True

            If String.IsNullOrEmpty(currentPassword) OrElse currentPassword.Length < requiredPasswordLength Then
                result = False
            Else
                Dim countAlphanumericCharacters As Integer = 0

                For Each character As Char In currentPassword
                    Select Case character
                        Case "!", "@", "$", "?", "_", "-"
                            countAlphanumericCharacters += 1

                        Case Else
                    End Select
                Next

                If nonAlphanumericCharacters > countAlphanumericCharacters Then
                    result = False
                End If
            End If

            Return result
        End Function

        Private Shared Function CreateRandomAlphanumericCharacters(charactersLength As Integer) As String
            Const allowedChars As String = "!@$?_-"

            Dim chars As Char() = New Char(charactersLength - 1) {}
            Dim newRandom As New Random()

            For index As Integer = 0 To charactersLength - 1
                chars(index) = allowedChars(newRandom.[Next](0, allowedChars.Length))
            Next

            Return New String(chars)
        End Function

#End Region

        '**%StrEncode: Password encryption routine
        '%StrEncode: Rutina de encriptamiento de password
        Public Shared Function StrEncode(ByVal s As String) As String
            Dim key As Integer
            Dim salt As Boolean
            Dim n As Integer
            Dim i As Integer
            Dim ss As String
            Dim k1 As Integer
            Dim k2 As Integer
            Dim k3 As Integer
            Dim k4 As Integer
            Dim t As Integer
            Dim result As String = String.Empty

            Static saltvalue As String
            If Trim(s) <> String.Empty Then

                key = 1234567890
                salt = False

                If salt Then
                    For i = 1 To 4
                        t = 100 * (1 + Asc(Mid(saltvalue, i, 1))) * Rnd() * (VB.Timer() + 1)
                        Mid(saltvalue, i, 1) = Chr(t Mod 256)
                    Next
                    s = Mid(saltvalue, 1, 2) & s & Mid(saltvalue, 3, 2)
                End If

                n = Len(s)
                ss = Space(n)
                Dim sn(n) As Integer

                k1 = 11 + (key Mod 233) : k2 = 7 + (key Mod 239)
                k3 = 5 + (key Mod 241) : k4 = 3 + (key Mod 251)

                For i = 1 To n : sn(i) = Asc(Mid(s, i, 1)) : Next

                For i = 2 To n : sn(i) = sn(i) Xor sn(i - 1) Xor ((k1 * sn(i - 1)) Mod 256) : Next
                For i = n - 1 To 1 Step -1 : sn(i) = sn(i) Xor sn(i + 1) Xor (k2 * sn(i + 1)) Mod 256 : Next
                For i = 3 To n : sn(i) = sn(i) Xor sn(i - 2) Xor (k3 * sn(i - 1)) Mod 256 : Next
                For i = n - 2 To 1 Step -1 : sn(i) = sn(i) Xor sn(i + 2) Xor (k4 * sn(i + 1)) Mod 256 : Next

                For i = 1 To n : Mid(ss, i, 1) = Chr(sn(i)) : Next
                result = ss
            End If
            Return result
        End Function

        '**%StrDecode: Password de-encryptment routine
        '%StrDecode: Rutina de des-encriptamiento de password
        Public Shared Function StrDecode(ByVal s As String) As String
            Dim key As Integer
            Dim salt As Boolean
            Dim n As Integer
            Dim i As Integer
            Dim ss As String
            Dim k1 As Integer
            Dim k2 As Integer
            Dim k3 As Integer
            Dim k4 As Integer
            Dim result As String = String.Empty

            If Trim(s) <> String.Empty Then

                key = 1234567890
                salt = False

                n = Len(s)
                ss = Space(n)
                Dim sn(n) As Integer

                k1 = 11 + (key Mod 233) : k2 = 7 + (key Mod 239)
                k3 = 5 + (key Mod 241) : k4 = 3 + (key Mod 251)

                For i = 1 To n : sn(i) = Asc(Mid(s, i, 1)) : Next

                For i = 1 To n - 2 : sn(i) = sn(i) Xor sn(i + 2) Xor (k4 * sn(i + 1)) Mod 256 : Next
                For i = n To 3 Step -1 : sn(i) = sn(i) Xor sn(i - 2) Xor (k3 * sn(i - 1)) Mod 256 : Next
                For i = 1 To n - 1 : sn(i) = sn(i) Xor sn(i + 1) Xor (k2 * sn(i + 1)) Mod 256 : Next
                For i = n To 2 Step -1 : sn(i) = sn(i) Xor sn(i - 1) Xor (k1 * sn(i - 1)) Mod 256 : Next

                For i = 1 To n : Mid(ss, i, 1) = Chr(sn(i)) : Next i

                If salt Then result = Mid(ss, 3, Len(ss) - 4) Else result = ss
            End If
            Return result
        End Function

    End Class

End Namespace

File: Common/InMotionGIT.Common/Helpers/AuthenticationHandler.cs
using System;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;

namespace InMotionGIT.Common.Helpers
{

    public class AuthenticationHandler
    {

        #region Random Password Methods

        public static string CreateAdministratorPassword(int requiredPasswordLength, int nonAlphanumericCharacters)
        {
            string newPassword = string.Empty;
            string characters = string.Empty;
            int passwordLength = requiredPasswordLength;

            if (nonAlphanumericCharacters > 0)
            {
                characters = CreateRandomAlphanumericCharacters(nonAlphanumericCharacters);
                passwordLength = requiredPasswordLength - characters.Length;
            }

            for (int index = 0, loopTo = passwordLength - 1; index <= loopTo; index++)
            {
                switch (index)
                {
                    case 0:
                        {
                            newPassword += "a";
                            break;
                        }
                    case 1:
                        {
                            newPassword += "d";
                            break;
                        }
                    case 2:
                        {
                            newPassword += "m";
                            break;
                        }
                    case 3:
                        {
                            newPassword += "i";
                            break;
                        }
                    case 4:
                        {
                            newPassword += "n";
                            break;
                        }

                    default:
                        {
                            newPassword += "1";
                            break;
                        }
                }
            }

            if (characters.Length > 0)
            {
                newPassword += characters;
            }

            return newPassword;
        }

        public static string CreateRandomPassword(int requiredPasswordLength, int nonAlphanumericCharacters)
        {
            string newPassword = string.Empty;
            string characters = string.Empty;
            int passwordLength = requiredPasswordLength;
            const string allowedChars = "abcdefghijkmnopqrstuvwxyzABCDEFGHJKLMNOPQRSTUVWXYZ0123456789";

            if (nonAlphanumericCharacters > 0)
            {
                characters = CreateRandomAlphanumericCharacters(nonAlphanumericCharacters);
                passwordLength = requiredPasswordLength - characters.Length;
            }

            var newRandom = new Random();

            for (int index = 0, loopTo = passwordLength - 1; index <= loopTo; index++)
                newPassword += Conversions.ToString(allowedChars[newRandom.Next(0, allowedChars.Length)]);

            if (characters.Length > 0)
            {
                newPassword += characters;
            }

            return newPassword;
        }

        public static bool IsCorrectPasswordBySettings(string currentPassword, int requiredPasswordLength, int nonAlphanumericCharacters)
        {
            bool result = true;

            if (string.IsNullOrEmpty(currentPassword) || currentPassword.Length < requiredPasswordLength)
            {
                result = false;
            }
            else
            {
                int countAlphanumericCharacters = 0;

                foreach (char character in currentPassword)
                {
                    switch (character)
                    {
                        case '!':
                        case '@':
                        case '$':
                        case '?':
                        case '_':
                        case '-':
                            {
                                countAlphanumericCharacters += 1;
                                break;
                            }

                        default:
                            {
                                break;
                            }
                    }
                }

                if (nonAlphanumericCharacters > countAlphanumericCharacters)
                {
                    result = false;
                }
            }

            return result;
        }

        private static string CreateRandomAlphanumericCharacters(int charactersLength)
        {
            const string allowedChars = "!@$?_-";

            char[] chars = new char[charactersLength];
            var newRandom = new Random();

            for (int index = 0, loopTo = charactersLength - 1; index <= loopTo; index++)
                chars[index] = allowedChars[newRandom.Next(0, allowedChars.Length)];

            return new string(chars);
        }

        private static string _StrEncode_saltvalue = default;

        #endregion

        // **%StrEncode: Password encryption routine
        // %StrEncode: Rutina de encriptamiento de password
        public static string StrEncode(string s)
        {
            int key;
            bool salt;
            int n;
            int i;
            string ss;
            int k1;
            int k2;
            int k3;
            int k4;
            int t;
            string result = string.Empty;
            if (!string.IsNullOrEmpty(Strings.Trim(s)))
            {

                key = 1234567890;
                salt = false;

                if (salt)
                {
                    for (i = 1; i <= 4; i++)
                    {
                        t = (int)Math.Round((double)(100 * (1 + Strings.Asc(Strings.Mid(_StrEncode_saltvalue, i, 1))) * VBMath.Rnd()) * (DateAndTime.Timer + 1d));
                        var midTmp = Conversions.ToString(Strings.Chr(t % 256));
                        StringType.MidStmtStr(ref _StrEncode_saltvalue, i, 1, midTmp);
                    }
                    s = Strings.Mid(_StrEncode_saltvalue, 1, 2) + s + Strings.Mid(_StrEncode_saltvalue, 3, 2);
                }

                n = Strings.Len(s);
                ss = Strings.Space(n);
                var sn = new int[n + 1];

                k1 = 11 + key % 233;
                k2 = 7 + key % 239;
                k3 = 5 + key % 241;
                k4 = 3 + key % 251;

                var loopTo = n;
                for (i = 1; i <= loopTo; i++)
                    sn[i] = Strings.Asc(Strings.Mid(s, i, 1));

                var loopTo1 = n;
                for (i = 2; i <= loopTo1; i++)
                    sn[i] = sn[i] ^ sn[i - 1] ^ k1 * sn[i - 1] % 256;
                for (i = n - 1; i >= 1; i -= 1)
                    sn[i] = sn[i] ^ sn[i + 1] ^ k2 * sn[i + 1] % 256;
                var loopTo2 = n;
                for (i = 3; i <= loopTo2; i++)
                    sn[i] = sn[i] ^ sn[i - 2] ^ k3 * sn[i - 1] % 256;
                for (i = n - 2; i >= 1; i -= 1)
                    sn[i] = sn[i] ^ sn[i + 2] ^ k4 * sn[i + 1] % 256;

                var loopTo3 = n;
                for (i = 1; i <= loopTo3; i++)
                {
                    var midTmp1 = Conversions.ToString(Strings.Chr(sn[i]));
                    StringType.MidStmtStr(ref ss, i, 1, midTmp1);
                }
                result = ss;
            }
            return result;
        }

        // **%StrDecode: Password de-encryptment routine
        // %StrDecode: Rutina de des-encriptamiento de password
        public static string StrDecode(string s)
        {
            int key;
            bool salt;
            int n;
            int i;
            string ss;
            int k1;
            int k2;
            int k3;
            int k4;
            string result = string.Empty;

            if (!string.IsNullOrEmpty(Strings.Trim(s)))
            {

                key = 1234567890;
                salt = false;

                n = Strings.Len(s);
                ss = Strings.Space(n);
                var sn = new int[n + 1];

                k1 = 11 + key % 233;
                k2 = 7 + key % 239;
                k3 = 5 + key % 241;
                k4 = 3 + key % 251;

                var loopTo = n;
                for (i = 1; i <= loopTo; i++)
                    sn[i] = Strings.Asc(Strings.Mid(s, i, 1));

                var loopTo1 = n - 2;
                for (i = 1; i <= loopTo1; i++)
                    sn[i] = sn[i] ^ sn[i + 2] ^ k4 * sn[i + 1] % 256;
                for (i = n; i >= 3; i -= 1)
                    sn[i] = sn[i] ^ sn[i - 2] ^ k3 * sn[i - 1] % 256;
                var loopTo2 = n - 1;
                for (i = 1; i <= loopTo2; i++)
                    sn[i] = sn[i] ^ sn[i + 1] ^ k2 * sn[i + 1] % 256;
                for (i = n; i >= 2; i -= 1)
                    sn[i] = sn[i] ^ sn[i - 1] ^ k1 * sn[i - 1] % 256;

                var loopTo3 = n;
                for (i = 1; i <= loopTo3; i++)
                {
                    var midTmp = Conversions.ToString(Strings.Chr(sn[i]));
                    StringType.MidStmtStr(ref ss, i, 1, midTmp);
                }

                if (salt)
                    result = Strings.Mid(ss, 3, Strings.Len(ss) - 4);
                else
                    result = ss;
            }
            return result;
        }

    }

}